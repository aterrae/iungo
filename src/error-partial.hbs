<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <title>Iungo - Error</title>
    <style>
    body {
        height: 100%;
        margin: 0;
        padding: 0;
        background-color: #59e0b3;
    }

    hgroup {
        margin: 0 auto;
        color: white;
        font-family: Arial, sans-serif;
    }

    hr {
        width: 80%;
        margin: 32px auto;
        border-top: 0;
        border-bottom: 2px dotted #ffffff;
    }

    .container {
        position: absolute;
        top: 50%;
        left: 50%;
        box-sizing: border-box;
        width: 100%;
        max-width: 768px;
        padding: 32px;
        transform: translate(-50%, -50%);
        text-align: center;
    }

    .title {
        margin: 24px 0 0 0;
        opacity: 1;
        text-shadow: 0 0 8px rgba(0, 0, 0, .3);
        font-size: 32px;
        font-weight: 900;
    }

    .filename {
        opacity: .8;
    }

    .message {
        margin: 0;
        color: #002e01;
    }

    .codeError {
        overflow: auto;
        box-sizing: border-box;
        max-width: 800px;
        max-height: 280px;
        margin: 32px auto 0 auto;
        padding: 20px 16px;
        text-align: left;
        border-radius: 8px;
        background: #002e01;
    }

    pre:empty {
        display: none;
    }

    pre::-webkit-scrollbar {
        width: 6px;
        height: 6px;
    }

    pre::-webkit-scrollbar-track {
        border-radius: 10px;
        background: rgba(0, 0, 0, .1);
    }

    pre::-webkit-scrollbar-thumb {
        border-radius: 10px;
        background: rgba(0, 0, 0, .2);
    }

    pre::-webkit-scrollbar-thumb:hover {
        background: rgba(0, 0, 0, .4);
    }

    pre::-webkit-scrollbar-thumb:active {
        background: rgba(0, 0, 0, .9);
    }
    </style>
</head>

<body>
    <div class="container">
        <hgroup>
            <h1 class="title">Iungo encountered an error!</h1>
            <hr/>
            <h2 class="filename">{{ error.fileName }}</h2>
            <h3 class="message">{{ error.message }}</h3>
            <div><pre class="codeError">{{{ error.codeError }}}</pre></div>
            <p>It's funny to join your code with Iungo</p>
        </hgroup>
    </div>
</body>

</html>
