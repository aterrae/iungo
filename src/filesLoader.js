import path from 'path';
import glob from 'glob';

function filesLoader(paths, pattern) {
    let files = [];

    paths = paths instanceof Array ? paths : [paths];

    for (let i in paths) {
        let fullPath = path.join(process.cwd(), paths[i], pattern);
        files = files.concat(glob.sync(fullPath));
    }

    return files;
}

export default filesLoader;
