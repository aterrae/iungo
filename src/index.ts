// Core utilities for Iungo
export { default as dataLoader } from './utils/dataLoader.js';
export { default as helpersLoader } from './utils/helpersLoader.js';
export { default as partialsLoader } from './utils/partialsLoader.js';
export { default as IungoError } from './utils/iungoError.js';
