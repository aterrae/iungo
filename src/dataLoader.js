import path from 'path';
import gutil from 'gulp-util';
import filesLoader from './filesLoader';

function dataLoader(data, paths) {
    let dataFiles = filesLoader(paths, '**/*.json');

    for (let i in dataFiles) {
        let name = path.basename(dataFiles[i]);
        let basename = path.basename(dataFiles[i], '.json');

        try {
            delete require.cache[require.resolve(dataFiles[i])];
            data[basename] = require(dataFiles[i]);
        } catch (error) {
            let errorData = {
                fileName: dataFiles[i],
                message: error.message
            }
            throw new gutil.PluginError('iungo', errorData);
        }
    }
}

export default dataLoader;
