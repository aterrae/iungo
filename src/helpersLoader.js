import Handlebars from "handlebars";
import PluginError from "plugin-error";
import path from "path";
import filesLoader from "./filesLoader";

function helpersLoader(paths) {
    let helperFiles = filesLoader(paths, "**/*.js");

    for (let i in helperFiles) {
        let basename = path.basename(helperFiles[i], ".js");

        try {
            if (Handlebars.helpers[basename]) {
                delete require.cache[require.resolve(helperFiles[i])];
                Handlebars.unregisterHelper(basename);
            }

            Handlebars.registerHelper(basename, require(helperFiles[i]));
        } catch (error) {
            let errorData = {
                fileName: helperFiles[i],
                message: "Error in helpers at " + error.loc.line + ":" + error.loc.column,
                stack: error.codeFrame
            }
            throw new PluginError("iungo", errorData);
        }
    }
}

export default helpersLoader;
