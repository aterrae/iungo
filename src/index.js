import Handlebars from 'handlebars';
import PluginError from 'plugin-error';
import through from 'through2';
import ansiHTML from 'ansi-html';
import dataLoader from './utils/dataLoader';
import helpersLoader from './utils/helpersLoader';
import partialsLoader from './utils/partialsLoader';
import errorPartial from './partials/error-partial.hbs';

class Iungo {
  constructor(opt) {
    this.opt = opt;
    this.data = {};

    ansiHTML.setColors({ reset: ['fff', '002e01'] });
  }

  render() {
    let stream = through.obj((chunk, encoding, callback) => {
      try {
        try {
          dataLoader(this.opt.data, this.data);
          partialsLoader(this.opt.partials);
          helpersLoader(this.opt.helpers);
        } catch (error) {
          throw new PluginError('iungo', error);
        }

        try {
          let page = Handlebars.compile(chunk.contents.toString());
          chunk.contents = new Buffer.from(page(this.data));
        } catch (error) {
          let errorData = {
            fileName: chunk.history[0],
            message: error.message,
          };
          throw new PluginError('iungo', errorData);
        }

        stream.push(chunk);
        callback();
      } catch (error) {
        this.data = {
          error: {
            fileName: error.fileName ? error.fileName : '',
            message: error.message ? error.message : '',
            codeError: error.stack ? ansiHTML(error.stack) : '',
          },
        };
        let page = Handlebars.compile(errorPartial);
        chunk.contents = new Buffer.from(page(this.data));

        stream.push(chunk);
        callback(error);
        stream.emit('end');
      }
    });

    return stream;
  }
}

let iungo;

export default (opt) => {
  /* istanbul ignore next */
  if (!iungo) {
    iungo = new Iungo(opt);
  }
  /* istanbul ignore next */
  return iungo.render();
};

export { Iungo };
