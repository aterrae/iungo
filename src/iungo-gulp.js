import Handlebars from 'handlebars';
import through from 'through2';
import ansiHTML from 'ansi-html';
import dataLoader from './utils/dataLoader';
import helpersLoader from './utils/helpersLoader';
import partialsLoader from './utils/partialsLoader';
import IungoError from './utils/iungoError';
import errorPartial from './partials/error-partial.hbs';

class IungoStream {
  constructor(options = {}) {
    this.options = Object.assign(
      {
        data: [],
        helpers: {},
        partials: [],
      },
      options,
    );

    this.data = {};

    ansiHTML.setColors({ reset: ['fff', '002e01'] });
  }

  render() {
    let stream = through.obj((chunk, encoding, callback) => {
      try {
        dataLoader(this.options.data, this.data);
        partialsLoader(this.options.partials);
        helpersLoader(this.options.helpers);

        try {
          let page = Handlebars.compile(chunk.contents.toString());
          chunk.contents = new Buffer.from(page(this.data));
        } catch (error) {
          throw new IungoError(error.message, chunk.history[0], error.stack);
        }

        stream.push(chunk);
        callback();
      } catch (error) {
        this.data = {
          error: {
            fileName: error.fileName ? error.fileName : '',
            message: error.message ? error.message : '',
            codeError: error.stack ? ansiHTML(error.stack) : '',
          },
        };
        let page = Handlebars.compile(errorPartial);
        chunk.contents = new Buffer.from(page(this.data));

        stream.push(chunk);
        callback(error);
        stream.emit('end');
      }
    });

    return stream;
  }
}

let iungo;

const IungoGulpPlugin = (options) => {
  /* istanbul ignore next */
  if (!iungo) {
    iungo = new IungoStream(options);
  }
  /* istanbul ignore next */
  return iungo.render();
};

export { IungoStream, IungoGulpPlugin };
