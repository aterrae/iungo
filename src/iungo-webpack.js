import Handlebars from 'handlebars';
import dataLoader from './utils/dataLoader';
import helpersLoader from './utils/helpersLoader';
import partialsLoader from './utils/partialsLoader';
import {
  getAbsolutePath,
  getBasename,
  getPathDiff,
  resolveGlob,
  readFile,
  registerDependency,
  getBasenameWithoutExt,
} from './utils/tools';
import IungoError from './utils/iungoError';

class IungoWebpackPlugin {
  constructor(options = {}) {
    this.options = Object.assign(
      {
        entry: null,
        output: null,
        data: [],
        helpers: {},
        partials: [],
        htmlWebpackPlugin: null,
        onBeforeInit: Function.prototype,
        onBeforeRegisterHelpers: Function.prototype,
        onBeforeRegisterPartials: Function.prototype,
        onBeforeCompile: Function.prototype,
        onBeforeRender: Function.prototype,
        onBeforeSave: Function.prototype,
        onDone: Function.prototype,
      },
      options,
    );

    this.options.entry = getAbsolutePath(this.options.entry);
    this.options.output = getAbsolutePath(this.options.output);

    this.options.onBeforeInit(Handlebars);
    this.data = {};
    this.fileDependencies = [];
    this.assetsGenerated = {};
    this.prevTimestamp = {};
    this.startTime = Date.now();
  }

  loadData() {
    dataLoader(this.options.data, this.data, this.fileDependencies);
  }

  loadHelpers() {
    helpersLoader(
      this.options.helpers,
      this.fileDependencies,
      this.options.onBeforeRegisterHelpers,
    );
  }

  loadPartials() {
    partialsLoader(
      this.options.partials,
      this.fileDependencies,
      this.options.onBeforeRegisterPartials,
    );
  }

  apply(compiler) {
    const compile = (compilation, callback) => {
      try {
        if (!this.dependenciesUpdated(compilation)) {
          callback();
          return;
        }

        this.loadData();
        this.loadHelpers();
        this.loadPartials();

        const paths = resolveGlob(`${this.options.entry}/**/*.html`);
        if (paths.length === 0) {
          callback();
          return;
        }
        paths.forEach((path) => {
          const content = readFile(path);
          this.compileFile(path, content, compilation);
        });
        callback();
      } catch (error) {
        compilation.errors.push(error);
        callback();
      }
    };

    const compileWithHtmlPlugin = (compilation, callback, data) => {
      try {
        if (!this.dependenciesUpdated(compilation)) {
          callback();
          return;
        }

        this.loadData();
        this.loadHelpers();
        this.loadPartials();

        let path = data.plugin.options.template.split('!').pop();
        let content = data.html;

        this.options.onBeforeSave = function(hb, hbTemplate) {
          data.html = hbTemplate;
        };

        this.compileFile(path, content, compilation);
        callback();
      } catch (error) {
        compilation.errors.push(error);
        callback();
      }
    };

    const registerDependencies = (compilation, callback) => {
      try {
        this.fileDependencies.forEach(
          compilation.fileDependencies.add,
          compilation.fileDependencies,
        );
        Object.keys(this.assetsGenerated).forEach((filename) => {
          compilation.assets[filename] = this.assetsGenerated[filename];
        });
        callback();
      } catch (error) {
        compilation.errors.push(
          new IungoError(`There was an error registering dependencies`),
        );
        callback();
      }
    };

    if (this.options.htmlWebpackPlugin) {
      compiler.hooks.compilation.tap('IungoWebpackPlugin', (compilation) => {
        this.options.htmlWebpackPlugin
          .getHooks(compilation)
          .afterTemplateExecution.tapAsync('IungoWebpackPlugin', (data, callback) => {
            compileWithHtmlPlugin(
              compilation,
              () => {
                registerDependencies(compilation, callback);
              },
              data,
            );
          });
      });
    } else {
      compiler.hooks.make.tapAsync('IungoWebpackPlugin', compile);
      compiler.hooks.emit.tapAsync('IungoWebpackPlugin', registerDependencies);
    }
  }

  dependenciesUpdated(compilation) {
    const timestampsSnapshot = compilation.fileTimestamps;
    const filePaths = Array.from(timestampsSnapshot.keys());

    const changedFiles = filePaths.filter((filePath) => {
      const prevTimestamp = this.prevTimestamp[filePath];
      const nextTimestamp = timestampsSnapshot.get(filePath);
      this.prevTimestamp[filePath] = nextTimestamp;

      return (prevTimestamp || this.startTime) < (nextTimestamp || Infinity);
    });

    if (changedFiles.length === 0) {
      return true;
    }

    for (let i = 0; i < changedFiles.length; i++) {
      if (this.fileDependencies.includes(changedFiles[i])) {
        return true;
      }
    }
    return false;
  }

  compileFile(path, content, compilation) {
    let targetPath;
    if (this.options.output) {
      targetPath = this.options.output.replace('[name]', getBasenameWithoutExt(path));
    } else {
      targetPath = `${compilation.compiler.outputPath}/${getBasename(path)}`;
    }

    const templateContent = this.options.onBeforeCompile(Handlebars, content) || content;
    const templateCompiled = Handlebars.compile(templateContent);
    const dataToRender = this.options.onBeforeRender(Handlebars, this.data) || this.data;
    let templateRendered = templateCompiled(dataToRender);
    templateRendered = this.options.onBeforeSave(Handlebars, templateRendered) || templateRendered;

    registerDependency(path, this.fileDependencies);

    if (!this.options.htmlWebpackPlugin) {
      if (targetPath.includes(compilation.compiler.outputPath)) {
        this.assetsGenerated[getPathDiff(targetPath, compilation.compiler.outputPath)] = {
          source: () => templateRendered,
          size: () => templateRendered.length,
        };
      } else {
        compilation.errors.push(
          new IungoError(`You can't set the output filepath outside the webpack dest folder`),
        );
      }
    }

    this.options.onDone(Handlebars, getPathDiff(targetPath, compilation.compiler.outputPath));
  }
}

export { IungoWebpackPlugin };
