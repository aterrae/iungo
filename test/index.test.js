import path from 'path';
import fs from 'fs';
import vfs from 'vinyl-fs';
import { Iungo } from '../src';

const FIXTURES = path.join(__dirname, 'fixtures/');

test('renders partial correctly', done => {
    expect.assertions(1);

    let i = new Iungo({
        partials: FIXTURES + 'partial/partials'
    });

    vfs.src(FIXTURES + 'partial/pages/index.html')
        .pipe(i.render())
        .pipe(vfs.dest(FIXTURES + 'partial/build'))
        .on('finish', () => {
            let build = fs.readFileSync(FIXTURES + 'partial/build/index.html').toString();
            expect(build).toMatchSnapshot();
            done();
        });
});

test('creates partial error correctly', done => {
    expect.assertions(1);

    let i = new Iungo({
        partials: FIXTURES + 'partial-error/partials'
    });

    vfs.src(FIXTURES + 'partial-error/pages/index.html')
        .pipe(i.render())
        .on('error', (error) => {
            expect(error.message).toMatchSnapshot();
            done();
        });
});

test('data is being passed correctly', done => {
    expect.assertions(2);

    let i = new Iungo({
        data: FIXTURES + 'data/data'
    });

    vfs.src(FIXTURES + 'data/pages/index.html')
        .pipe(i.render())
        .pipe(vfs.dest(FIXTURES + 'data/build'))
        .on('finish', () => {
            expect(i.data).toEqual({
                strings: {
                    name: "Aterrae",
                    location: "Rubano"
                }
            });
            let build = fs.readFileSync(FIXTURES + 'data/build/index.html').toString();
            expect(build).toMatchSnapshot();
            done();
        });
});

test('creates data error correctly', done => {
    expect.assertions(1);

    let i = new Iungo({
        data: FIXTURES + 'data-error/data'
    });

    vfs.src(FIXTURES + 'data-error/pages/index.html')
        .pipe(i.render())
        .on('error', (error) => {
            expect(error.message).toMatchSnapshot();
            done();
        });
});

test('uses helpers correctly', done => {
    expect.assertions(1);

    let i = new Iungo({
        helpers: FIXTURES + 'helper/helpers'
    });

    vfs.src(FIXTURES + 'helper/pages/index.html')
        .pipe(i.render())
        .pipe(vfs.dest(FIXTURES + 'helper/build'))
        .on('finish', () => {
            let build = fs.readFileSync(FIXTURES + 'helper/build/index.html').toString();
            expect(build).toMatchSnapshot();
            done();
        });
});

test('creates helpers error correctly', done => {
    expect.assertions(1);

    let i = new Iungo({
        helpers: FIXTURES + 'helper-error/helpers'
    });

    vfs.src(FIXTURES + 'helper-error/pages/index.html')
        .pipe(i.render())
        .on('error', (error) => {
            expect(error.message).toMatchSnapshot();
            done();
        });
});
