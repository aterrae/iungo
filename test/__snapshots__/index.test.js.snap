// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates data error correctly 1`] = `"Unexpected token } in JSON at position 47"`;

exports[`creates helpers error correctly 1`] = `"Error in helpers at 2:34"`;

exports[`creates partial error correctly 1`] = `"The partial bodi could not be found"`;

exports[`data is being passed correctly 1`] = `
"<html>
    <body>
        <p>Aterrae</p>
        <p>Rubano</p>
    </body>
</html>
"
`;

exports[`renders partial correctly 1`] = `
"<html>
    <head>
        <title>Title</title>
        
    </head>
    <body>
        <p>Body!</p>
        
    </body>
</html>
"
`;

exports[`uses helpers correctly 1`] = `
"<html>
    <body>
        <p>I&#x27;m a Iungo helper!</p>
    </body>
</html>
"
`;
